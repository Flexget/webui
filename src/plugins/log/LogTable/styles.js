import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { Table as RVTable } from 'react-virtualized';

const color = (theme, hex) => css`
  ${theme.palette.type === 'light'
    ? css`
        background-color: ${hex};
      `
    : css`
        color: ${hex};
      `}
`;

export const Table = styled(RVTable)`
  font-size: 1rem;

  & .error {
    ${({ theme }) => color(theme, '#f2dede')};
  }

  & .critical {
    ${({ theme }) => color(theme, '#f2dede')};
  }

  & .warning {
    ${({ theme }) => color(theme, '#fcf8e3')};
  }
`;
