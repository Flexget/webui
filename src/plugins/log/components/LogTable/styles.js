import styled from '@emotion/styled';
import { Table as RVTable } from 'react-virtualized';

export const Table = styled(RVTable)`
  font-size: 1rem;

  & .error {
    background-color: #f2dede;
  }

  & .critical {
    background-color: #f2dede;
  }

  & .warning {
    background-color: #fcf8e3;
  }
`;
