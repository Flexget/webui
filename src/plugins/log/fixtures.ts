import { random, date } from 'faker';
import { LogMessage, LogLevel } from './types';

export const makeLogMessage = (logLevel = LogLevel.Warning): LogMessage => ({
  timestamp: date.past().toISOString(),
  message: random.words(),
  task: random.words(),
  logLevel,
  plugin: random.words(),
});
