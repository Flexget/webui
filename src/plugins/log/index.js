import { createAsyncComponent } from 'utils/loading';
import registry from 'core/registry';
import saga from './state/saga';
import reducer from './state/reducer';

export default () => registry.registerPlugin('log', {
  component: createAsyncComponent(() => import('plugins/log/Log')),
  routeDisplayName: 'Log',
  routeIcon: 'book',
  reducer,
  saga,
});
