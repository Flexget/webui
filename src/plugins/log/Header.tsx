import React, {
  FC,
  useState,
  useCallback,
  MouseEvent,
  useMemo,
  ComponentType,
  useEffect,
} from 'react';
import { useFormikContext } from 'formik';
import {
  Theme,
  Typography,
  IconButton,
  Menu,
  MenuItem,
  ListItemIcon,
  CircularProgress,
} from '@material-ui/core';
import { css } from '@emotion/core';
import { Spacer } from 'common/styles';
import TextField from 'common/inputs/formik/TextField';
import { PlayArrow, Stop, FilterList, MoreVert, ClearAll } from '@material-ui/icons';
import { ReadyState } from 'core/api';
import { useDebounce } from 'utils/hooks';
import { Options } from './types';

export const wrapper = (theme: Theme) => css`
  display: block;
  ${theme.breakpoints.up('sm')} {
    display: flex;
    align-items: baseline;
  }
`;

export const filterWrapper = css`
  display: flex;
  align-items: center;
`;

export const filterField = css`
  margin-left: 1rem;
  margin-right: 1rem;
`;

interface Props {
  readyState: ReadyState;
  connect: () => void;
  disconnect: () => void;
  clear: () => void;
}

interface StateOptions {
  onClick?: () => void;
  Icon: ComponentType;
  label: string;
  heading: string;
  disabled?: boolean;
}

const Header: FC<Props> = ({ readyState, connect, disconnect, clear }) => {
  const [anchorEl, setAnchorEl] = useState<HTMLButtonElement>();
  const helperText = 'Supports operators and, or, (), and "str"';

  const stateMap: Record<ReadyState, StateOptions> = useMemo(
    () => ({
      [ReadyState.Open]: {
        onClick: disconnect,
        Icon: Stop,
        label: 'Stop',
        heading: 'Streaming',
      },
      [ReadyState.Connecting]: {
        label: 'Start',
        Icon: CircularProgress,
        heading: 'Connecting',
        disabled: true,
      },
      [ReadyState.Closed]: {
        onClick: connect,
        Icon: PlayArrow,
        label: 'Start',
        heading: 'Disconnected',
      },
    }),
    [connect, disconnect],
  );

  const { onClick, Icon, heading, label, disabled = false } = stateMap[readyState];

  const handleMenuClose = useCallback(() => setAnchorEl(undefined), []);
  const handleMenuOpen = useCallback(
    (event: MouseEvent<HTMLButtonElement>) => setAnchorEl(event.currentTarget),
    [],
  );

  const { values, submitForm } = useFormikContext<Options>();

  const debouncedValues = useDebounce(values);

  useEffect(() => {
    submitForm();
  }, [...Object.values(debouncedValues), submitForm]);

  return (
    <div css={wrapper}>
      <div>
        <Typography variant="h6">Server Log</Typography>
        <Typography variant="subtitle1" color="textSecondary">
          {heading}
        </Typography>
      </div>
      <Spacer />
      <div css={filterWrapper}>
        <FilterList />
        <TextField
          css={filterField}
          id="query"
          name="query"
          label="Filter"
          helperText={helperText}
        />
        <IconButton onClick={handleMenuOpen}>
          <MoreVert />
        </IconButton>
      </div>
      <Menu id="log-menu" anchorEl={anchorEl} open={!!anchorEl} onClose={handleMenuClose}>
        <MenuItem>
          <TextField id="lines" name="lines" label="Max Lines" type="number" />
        </MenuItem>
        <MenuItem onClick={clear}>
          <ListItemIcon>
            <ClearAll />
          </ListItemIcon>
          Clear
        </MenuItem>
        <MenuItem onClick={onClick} disabled={disabled}>
          <ListItemIcon>
            <Icon />
          </ListItemIcon>
          {label}
        </MenuItem>
      </Menu>
    </div>
  );
};

export default Header;
