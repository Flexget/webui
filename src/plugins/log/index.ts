import { createAsyncComponent } from 'utils/loading';
import { registerPlugin } from 'core/plugins/registry';
import ListAltIcon from '@material-ui/icons/ListAlt';

export default () =>
  registerPlugin('/log', {
    component: createAsyncComponent(() =>
      import(
        /* webpackChunkName: 'LogPlugin' */
        'plugins/log/Log'
      ),
    ),
    routeDisplayName: 'Log',
    routeIcon: ListAltIcon,
  });
