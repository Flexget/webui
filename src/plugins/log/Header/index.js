import { connect } from 'react-redux';
import actions from 'plugins/log/state/actions';
import Header from './Header';

export function mapStateToProps({ log }) {
  return {
    connected: log.connected,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    start: payload => dispatch(actions.connect.request(payload)),
    stop: () => dispatch(actions.disconnect()),
    clearLogs: () => dispatch(actions.clear()),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Header);
export { Header };
