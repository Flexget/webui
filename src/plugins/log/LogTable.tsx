import React, { FC } from 'react';
import { AutoSizer, Column, Table } from 'react-virtualized';
import 'react-virtualized/styles.css';
import { css, ClassNames } from '@emotion/core';
import { Theme, useTheme } from '@material-ui/core';
import { LogMessage, LogLevel } from 'plugins/log/types';

const color = (theme: Theme, hex: string) =>
  theme.palette.type === 'light'
    ? css`
        background-color: ${hex};
      `
    : css`
        color: ${hex};
      `;

const table = css`
  font-size: 1rem;
`;

interface Props {
  messages: LogMessage[];
}

const LogTable: FC<Props> = ({ messages }) => {
  const theme = useTheme();

  return (
    <ClassNames>
      {({ css: cssString }) => {
        const cssMap = {
          [LogLevel.Error]: cssString(color(theme, '#f2dede')),
          [LogLevel.Critical]: cssString(color(theme, '#f2dede')),
          [LogLevel.Warning]: cssString(color(theme, '#fcf8e3')),
        };

        return (
          <AutoSizer>
            {({ height, width }) => {
              let actualWidth = width;
              let actualHeight = height;
              if (process.env.NODE_ENV === 'test') {
                actualWidth = window.innerWidth;
                actualHeight = window.innerHeight;
              }
              return (
                <Table
                  css={table}
                  rowCount={messages.length}
                  rowHeight={20}
                  headerHeight={50}
                  width={actualWidth}
                  height={actualHeight}
                  rowGetter={({ index }) => messages[index]}
                  rowClassName={({ index }) => cssMap[messages[index]?.logLevel]}
                >
                  <Column label="Time" dataKey="timestamp" width={100} />
                  <Column label="Level" dataKey="logLevel" width={100} />
                  <Column label="Plugin" dataKey="plugin" width={100} />
                  <Column label="Task" dataKey="task" width={100} />
                  <Column label="Message" dataKey="message" width={100} flexGrow={1} />
                </Table>
              );
            }}
          </AutoSizer>
        );
      }}
    </ClassNames>
  );
};

export default LogTable;
