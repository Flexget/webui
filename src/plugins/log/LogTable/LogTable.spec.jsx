import React from 'react';
import { shallow } from 'enzyme';
import { themed } from 'utils/tests';
import { mapStateToProps, LogTable } from 'plugins/log/LogTable';

describe('plugins/log/components/LogTable', () => {
  describe('LogTable', () => {
    it('renders correctly', () => {
      const wrapper = shallow(themed(<LogTable messages={[]} />));
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('mapStateToProps', () => {
    it('should return the right stuff', () => {
      expect(
        mapStateToProps({
          log: {
            messages: [],
          },
        }),
      ).toMatchSnapshot();
    });
  });
});
