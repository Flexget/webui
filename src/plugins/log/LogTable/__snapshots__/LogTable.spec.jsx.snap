// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/log/components/LogTable LogTable renders correctly 1`] = `
<Provider>
  <Theme>
    <LogTable
      messages={Array []}
    />
  </Theme>
</Provider>
`;

exports[`plugins/log/components/LogTable mapStateToProps should return the right stuff 1`] = `
Object {
  "messages": Array [],
}
`;
